lib/streamagg: make flush faster via concurrent output relabling application

BenNF opened the pull request:
### Describe Your Changes

With stream aggregations that output an extremely large number of samples and apply output relabeling the sequential application of relabeling slows down the flush output.

This can result in flushes taking longer than the aggregation interval and therefore producing unexpected gaps in stream aggregation output. 

This work currently does not scale well with additional CPUs. This PR adds some simple concurrency which shows improved performance in my testing.


p90 stream aggregation flush time drops by nearly 50% in our setup.
<img width="1433" height="634" alt="Screenshot 2025-08-11 at 12 01 21 PM" src="https://github.com/user-attachments/assets/f30493f9-fb8c-48a4-98cf-867cdfab41ca" />

And while this PR does remove some reused objects we don't see any worse GC performance, however our GC performance is already high cost so this may not be representative. 
<img width="1440" height="651" alt="Screenshot 2025-08-11 at 12 01 34 PM" src="https://github.com/user-attachments/assets/a5078265-6024-4cc7-8187-e5a33aeeb1bc" />


### Checklist

The following checks are **mandatory**:
- [x] My change adheres to [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/victoriametrics/contributing/#pull-request-checklist).
- [x] My change adheres to [VictoriaMetrics development goals](https://docs.victoriametrics.com/victoriametrics/goals/).

BenNF committed:
lib/streamagg: make flushSeries faster via concurrent output relabling

rework flushSeries to use worker pool for large batches of output relabeling

AndrewChubatiuk commented on lib/streamaggr/streamaggr.go:
```diff
@@ -1210,26 +1211,37 @@ func (ctx *flushCtx) flushSeries() {
+		wg.Add(1)
+		go func(ts *prompb.TimeSeries) {
+			defer wg.Done()
+			dstLabels := make([]prompb.Label, 0, len(ts.Labels))
```
why don't you use `promutil.GetLabels()` here?

BenNF replied:
No real strong reason, I'll switch it over

zekker6 commented on lib/streamaggr/streamaggr.go:
```diff
@@ -1210,26 +1211,44 @@ func (ctx *flushCtx) flushSeries() {
-		if len(dstLabels) == dstLabelsLen {
-			// The metric has been deleted by the relabeling
-			continue
-		}
-		ts.Labels = dstLabels[dstLabelsLen:]
-		dst = append(dst, ts)
+		rSem <- struct{}{}
+		wg.Add(1)
+		go func(ts *prompb.TimeSeries) {
+			defer wg.Done()
+
+			auxLabels := promutil.GetLabels()
+			dstLabels := auxLabels.Labels[:0]
+			dstLabelsLen := len(dstLabels)
+
+			dstLabels = append(dstLabels, ts.Labels...)
+			dstLabels = outputRelabeling.Apply(dstLabels, dstLabelsLen)
+			if len(dstLabels) == dstLabelsLen {
+				// The metric has been deleted by the relabeling
+				return
+			}
+
+			ts.Labels = append([]prompb.Label(nil), dstLabels[dstLabelsLen:]...)
+			auxLabels.Labels = dstLabels
+			promutil.PutLabels(auxLabels)
+			dstCh <- ts
+			<-rSem
+		}(&ts)
+	}
```
I see a few issues with this approach currently:
- `rSem` does is not getting the "slot" back in case of early return due to tss being removed. This might make block execution of concurrent goroutines.
- unsafe usage of auxLabels - current code calls `ctx.pushFunc` before releasing `auxLabels` back to pool, this is safe because output will be flushed before memory is reclaimed. An updated version uses labels from `auxLabels` which might be reclaimed by GC, so labels will become corrupted.
- optimisation wise - current uses a single `auxLabels.Labels` slice in order to capture labels for all `tss`. Each `ts` is using a subslice of the main slice in order to reference a specific set of labels. This allows to reduce amount of smaller allocations for each `ts` and uses a single large slice instead. An updated version will create get a labels for each `ts` so the optimisation is a lot less effective.


I would suggest to change an approach here from spawning a goroutine for each `ts`in `tss` to spin up a worker pool limited by `cgroup.AvailableCPUs`. Each worker in pool will be able to reuse `auxLabels` when applying relabeling. Note that workers' `auxLabels` must be returned back to pool only after `pushFunc` will be called in order to make it safe.  Also, it most likely makes sense to spin up workers pool only in case `len(tss) > cgroup.AvailableCPUs` as otherwise cost of spinning up worker pool might be higher than benefits it gives.

Please, let me know if it is not clear and I'll elaborate on specific topics in more details.

Commit:
return reference to output

Commit:
lib/streamaggr rework concurrent relabeling

rework concurrent relabeling to a batching approach
rather than a channel based apporach due to GC issues
during testing

BenNF commented:
@zekker6 I've reworked this quite a lot and I'd love your thoughts. 

I did initially go with a simple channel based worker pool approach(see [here](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/9571/commits/72ea2871246feea694ebe798f60ad255de12c3a0)), but I ended up with worse performance due to GC time. 

So I changed approaches and went with this pre-batched approach trying very hard to reduce allocations. This is showing very similar, or slightly better,  GC time in my testing to the current implementation. 

However I'd like to call out that this is _only_ actually a performance improvement in the case of single stream aggregations with very large outputs, which was the initial setup I optimized for. 

We've since split that single large stream aggregation into many smaller stream aggregations running on the same agent and no longer see any significant speed up improvement, likely since each aggregation is executing in parallel already saturate the cpu, so additional parallelism _within_ the aggregation flush doesn't add value. 

This optimization is still effective in the "N aggregations with large outputs on M cpu cores, N<M" case, but I'll leave it up to you to make the call on weather the complexity is worth it for just that case.
